drivers/gpio/gpio_gecko.c

```diff
@@ -0,0 +1,425 @@
+/*
+ * Copyright (c) 2017, Christian Taedcke
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <gpio.h>
+#include <soc.h>
+#include <em_gpio.h>
+
+#include "gpio_utils.h"
+
+#define member_size(type, member) sizeof(((type *)0)->member)
+#define NUMBER_OF_PORTS (member_size(GPIO_TypeDef, P) / \
+			 member_size(GPIO_TypeDef, P[0]))
+
+struct gpio_gecko_common_config {
+};
+
+struct gpio_gecko_common_data {
+	/* a list of all ports */
+	struct device *ports[NUMBER_OF_PORTS];
+	size_t count;
+};
+
+struct gpio_gecko_config {
+	GPIO_P_TypeDef *gpio_base;
+	GPIO_Port_TypeDef gpio_index;
+};
+
+struct gpio_gecko_data {
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* pin callback routine enable flags, by pin number */
+	u32_t pin_callback_enables;
+};
+
+static inline void gpio_gecko_add_port(struct gpio_gecko_common_data *data,
+				       struct device *dev)
+{
+	__ASSERT(dev, "No port device!");
+	data->ports[data->count++] = dev;
+}
+
+
+static int gpio_gecko_configure(struct device *dev,
+				int access_op, u32_t pin, int flags)
+{
+	const struct gpio_gecko_config *config = dev->config->config_info;
+	GPIO_P_TypeDef *gpio_base = config->gpio_base;
+	GPIO_Port_TypeDef gpio_index = config->gpio_index;
+	GPIO_Mode_TypeDef mode;
+	unsigned int out = 0;
+	bool interrupt_rising_edge = false;
+	bool interrupt_falling_edge = false;
+	bool enable_interrupt = false;
+
+	/* Check for an invalid pin configuration */
+	if ((flags & GPIO_INT) && (flags & GPIO_DIR_OUT)) {
+		return -EINVAL;
+	}
+
+	/* Interrupt on static level is not supported by the hardware */
+	if ((flags & GPIO_INT) && ((flags & GPIO_INT_EDGE) == GPIO_INT_LEVEL)) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
+
+		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+			mode = gpioModeInputPull;
+			out = 1; /* pull-up*/
+		} else if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_DOWN) {
+			mode = gpioModeInputPull;
+			/* out = 0 means pull-down*/
+		}
+
+		mode = gpioModeInput;
+	} else { /* GPIO_DIR_OUT */
+		mode = gpioModePushPull;
+	}
+	/* The flags contain options that require touching registers in the
+	 * GPIO module and the corresponding PORT module.
+	 *
+	 * Start with the GPIO module and set up the pin direction register.
+	 * 0 - pin is input, 1 - pin is output
+	 */
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		GPIO_PinModeSet(gpio_index, pin, mode, out);
+	} else {	/* GPIO_ACCESS_BY_PORT */
+		gpio_base->MODEL = (mode << 28) | (mode << 24) | (mode << 20)
```

Comment:
what's going on here?

Reply:
This configure each pin of the given port according to the mode. 4 Bits contain the configuration for one pin, see https://www.silabs.com/documents/public/reference-manuals/EFM32WG-RM.pdf pg. 982.

Reply:
create macros that are more readable for these something like GECKO_MODE_<X/Y/Z> and add the datasheet reference as comment.

Reply:
Would this be better?

~~~
/*
 * Macros to set the gpio mode registers
 * 
 * See https://www.silabs.com/documents/public/reference-manuals/EFM32WG-RM.pdf
 * page 972 and 982.
 */
/**
 * @brief Create the value to set the GPIO MODEL register
 * @param[in] pin The index of the pin. Valid values are 0..7.
 * @param[in] mode The mode that should be set.
 * @return The value that can be set into the GPIO MODEL register.
 */
#define GECKO_GPIO_MODEL(pin, mode) (mode << (pin * 4))

/**
 * @brief Create the value to set the GPIO MODEH register
 * @param[in] pin The index of the pin. Valid values are 8..15.
 * @param[in] mode The mode that should be set.
 * @return The value that can be set into the GPIO MODEH register.
 */
#define GECKO_GPIO_MODEH(pin, mode) (mode << ((pin - 8) * 4))

gpio_base->MODEL = GECKO_GPIO_MODEL(7, mode) | GECKO_GPIO_MODEL(6, mode)
	| GECKO_GPIO_MODEL(5, mode) | GECKO_GPIO_MODEL(4, mode)
	| GECKO_GPIO_MODEL(3, mode) | GECKO_GPIO_MODEL(2, mode)
	| GECKO_GPIO_MODEL(1, mode) | GECKO_GPIO_MODEL(0, mode);
gpio_base->MODEH = GECKO_GPIO_MODEH(15, mode) | GECKO_GPIO_MODEH(14, mode)
	| GECKO_GPIO_MODEH(13, mode) | GECKO_GPIO_MODEH(12, mode)
	| GECKO_GPIO_MODEH(11, mode) | GECKO_GPIO_MODEH(10, mode)
	| GECKO_GPIO_MODEH(9, mode) | GECKO_GPIO_MODEH(8, mode);
~~~

Reply:
fine by me, it is not mandatory to use doxygen format there actually.

Reply:
ok, done